### PR TITLE
ci: Retry on timeout

### PR DESCRIPTION
**Pull request by ko3n1g**
# What does this PR do ?

This PR improves resiliency by auto-retrying tests that ran into timeouts. We retry two times (3 runs in total) and only retry on timeout violation. In the run overview, retries can be seen via `Annotation`. 

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/core/test_exp_manager.py [hunk 1]
```diff
@@ -151,6 +151,7 @@ def test_omegaconf(self):
             exp_manager(pl.Trainer(accelerator='cpu'), {"unused": 1})
 
     @pytest.mark.unit
+    @pytest.mark.pleasefixme
```
Comment (chtruong814) [inline tests/core/test_exp_manager.py:154]: Are these new tests that are inconsistently failing now?

**Review by chtruong814 — DISMISSED** [review]
Great idea. I'm glad we're limiting it to retrying on timeouts only rather than everything.

Just had one clarification question but overall looks good to me.

File: tests/core/test_exp_manager.py (hunk 1 above)
Comment (ko3n1g) [reply tests/core/test_exp_manager.py:154]: These have been failing for a long time unfortunately, but in the context of this PR I finally tackle these

File: .github/workflows/cicd-main.yml [hunk 2]
```diff
@@ -183,10 +183,10 @@ jobs:
       SCRIPT: |
         NEMO_NUMBA_MINVER=0.53 pytest tests/collections/tts -m "not pleasefixme" --with_downloads --cov-report=term  --cov-branch --cov-report=xml --cov=nemo
 
-  OPTIONAL_L0_Unit_Tests_GPU_Core:
+  L0_Unit_Tests_GPU_Core:
```
Comment (chtruong814) [inline .github/workflows/cicd-main.yml:186]: Should we also remove the IS_OPTIONAL flag here like the other tests you modified?

Comment (ko3n1g) [reply .github/workflows/cicd-main.yml:186]: good catch!
